### ISSUE TITLE
Spec bug for lifted short-circuiting Logical Operators

### ISSUE THREAD

**Issue by gafter** [dotnet/vblang#478]
In 11.17.1 the VB specification says

> In the lifted form of the `AndAlso` and `OrElse` operators, if the first operand was a null `Boolean?`, then the second operand is evaluated but the result is always a null `Boolean?`.

But this is not correct or what is implemented (at least for the types `Boolean?`).  It should say

> In the lifted form of the `AndAlso` and `OrElse` operators, if the first operand was a null `Boolean?`, then the second operand is evaluated and is the result.

I'm not sure how this affects these operators in the presence of a user-declared type with `IsTrue` `IsFalse`, `And` and `Or`.

**Comment by jrmoreno1** [comment]
That’s not quite right. It should say:

In the lifted form of OrElse operator, if the first operand was a null Boolean?, then the second operand is evaluated and is the result.

VB uses sql like nulls, where null is neither true or false.  With short circuiting AndAlso can never be true if the first operand is null because null and true/false/null is going to be null.  OrElse can be true of course, so it needs to evaluated if the first operand is null.

**Comment by jrmoreno1** [comment]
@gafter 

Program to show that the second operand isn’t evaluated if the first operand is null `Boolean?`, and the operator is `AndAlso`:

    Imports System

    Public Module Module1

        Public Sub Main()
            dim example as boolean? = nothing
            dim result  = example AndAlso ThrowEx()
            console.WriteLine("side effect avoided")
        End Sub
        Public Function ThrowEx()
            throw new exception("side effect hit")
        End Function
    End Module

**Comment by gafter** [comment]
@AlekseyTs Does this relate to your recent VB compiler bug fix?

**Comment by AlekseyTs** [comment]
> Does this relate to your recent VB compiler bug fix?

No. The example above is not an example of lifted short-circuiting Logical Operator. It is an example of late-bound short-circuiting Logical Operator. That one doesn't utilize three-valued logic. It is sufficient to change return type of ```ThrowEx``` function to Boolean (instead of Object) in order to make it an example of lifted short-circuiting Logical Operator and observe the "side effect hit".

**Comment by rskar-git** [comment]
@jrmoreno1

> With short circuiting AndAlso can never be true if the first operand is null because null and true/false/null is going to be null.

Not quite right, `(null AndAlso False)` should result in False if we're being faithful to three-valued logic (https://en.wikipedia.org/wiki/Three-valued_logic).

In your example code, `example AndAlso ThrowEx()` should have resulted in `ThrowEx()` getting evaluated, so if VB isn't doing that now, really it should have been. AndAlso should short-circuit only on the first parameter being False; and for OrElse, first parameter being True. Null Boolean? may make this seem trickier, but really it shouldn't - that general rule should still work: Both AndAlso and OrElse should continue evaluating the second parameter if the first is null.

I suppose for further reflection on this it may be worth wading through https://github.com/dotnet/csharplang/issues/871. For whatever the reason, there are folks there who are seeing this bool? and && and || business as a problematic "mapping null to false" thing.

**Comment by jrmoreno1** [comment]
@rskar-git : You’re right, null And False should logically be False.  Also, it looks like the behavior is different depending upon the return type of the ThrowEx function (sorry, typing on an iPad, makes it harder to get all of the details right). When the type is Object, the function doesn’t get called, when it’s a bool it does.
